### PR TITLE
[GEOS-7353] Make GetFeatureInfo HTML template loading thread-safe.

### PR DESCRIPTION
**Pull request by sikeoka**
Fix for:
[GEOS-7353] GetFeatureInfo HTML template loading is not thread-safe

Can be backported to 2.8.x and 2.7.x.

### CODE REVIEW

File: src/main/src/main/java/org/geoserver/template/GeoServerTemplateLoader.java [hunk 1]
```diff
@@ -83,7 +83,7 @@
      * Feature type directory to load template against. Its presence is mutually
      * exclusive with coverageName
      */
-    ResourceInfo resource;
+    protected ResourceInfo resource;
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/template/GeoServerTemplateLoader.java]: Is this change actually used anywhere?

File: src/wms/src/main/java/org/geoserver/wms/featureinfo/HTMLFeatureInfoOutputFormat.java [hunk 2]
```diff
@@ -68,7 +68,7 @@ public TemplateModel wrap(Object object) throws TemplateModelException {
         });
     }
 
-    GeoServerTemplateLoader templateLoader;
+    volatile GeoServerTemplateLoader templateLoader;
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/featureinfo/HTMLFeatureInfoOutputFormat.java]: Normally I see volatile being used along with a double checked locking pattern, but here it does not seem the case. Wasn't moving the loader init in the synchronized block enough?

File: src/main/src/main/java/org/geoserver/template/GeoServerTemplateLoader.java (hunk 1 above)
Comment (aaime) [inline src/main/src/main/java/org/geoserver/template/GeoServerTemplateLoader.java]: Ah, seems to be used in the tests

File: src/wms/src/main/java/org/geoserver/wms/featureinfo/HTMLFeatureInfoOutputFormat.java (hunk 2 above)
Comment (sikeoka) [inline src/wms/src/main/java/org/geoserver/wms/featureinfo/HTMLFeatureInfoOutputFormat.java]: Yes, you are correct.  The volatile modifier is not necessary now that it is in the synchronized block.  I will remove it from the pull request.

**Comment by sikeoka** [comment]
I updated the pull request to remove the unnecessary volatile modifier.

**Comment by aaime** [comment]
Manually merged to squash the two commits into one, and then backported to 2.8.x and 2.7.x

**Pull request closed without merging**